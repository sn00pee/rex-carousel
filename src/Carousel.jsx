import React, { Component } from 'react';
import { node, shape, number, string, arrayOf, func, bool } from 'prop-types';
import './Carousel.scss';
import 'swiper/dist/css/swiper.css';
import Swiper from 'react-id-swiper';
import SwiperModule from 'swiper';
import CarouselTabs from './CarouselTabs';
import CarouselItem from './CarouselItem';
import 'rex-icons';

class Carousel extends Component {
  constructor(props) {
    super(props);

    const { items, isTabActive } = this.props;

    this.state = {
      swiper: null,
      activeIndex: 0,
      visibleTabsNum: 0,
      hiddenNextTabsNum: 0,
      hiddenPrevTabsNum: 0,
      currentTabsPos: 0,
      itemsLength: items.length - 2,
      isTabActive,
    };

    this.goToIndex = this.goToIndex.bind(this);
    this.onSlideChange = this.onSlideChange.bind(this);
  }

  componentDidMount() {
    this.calculateVisibleTabs();
  }

  onSlideChange() {
    const { swiper, activeIndex } = this.state;
    const swiperActiveIndex = swiper.activeIndex;
    const tabsInnnerContainer = document.getElementsByClassName(
      'carousel-tabs'
    )[0];

    if (!tabsInnnerContainer) return;

    if (swiperActiveIndex === activeIndex + 1) {
      this.showNextTab();
    } else if (swiperActiveIndex === activeIndex - 1) {
      this.showPrevTab();
    } else {
      this.showIndexTab(swiperActiveIndex);
    }

    this.setState({
      activeIndex: swiperActiveIndex,
    });
  }

  calculateVisibleTabs() {
    const tabsContainer = document.getElementsByClassName(
      'carousel-tabs-inner'
    )[0];

    if (!tabsContainer) return;

    const { itemsLength } = this.state;
    const tabsContainerWidth = tabsContainer.clientWidth;
    const tabItemWidth = document.getElementsByClassName('carousel-tab')[0]
      .clientWidth;
    const visibleTabsNum = Math.floor(tabsContainerWidth / tabItemWidth);
    const hiddenNextTabsNum = itemsLength - visibleTabsNum;

    this.setState({
      visibleTabsNum,
      hiddenNextTabsNum,
    });
  }

  showNextTab() {
    const {
      visibleTabsNum,
      hiddenPrevTabsNum,
      hiddenNextTabsNum,
      currentTabsPos,
      activeIndex,
    } = this.state;

    // At showing hidden next tabs
    if (activeIndex > visibleTabsNum - 2) {
      const tabsInnnerContainer = document.getElementsByClassName(
        'carousel-tabs'
      )[0];
      const tabItemWidth = document.getElementsByClassName('carousel-tab')[0]
        .clientWidth;
      let nextTabsPos = currentTabsPos - tabItemWidth;

      // At last tab, calculate hidden left width and only move the length
      if (hiddenNextTabsNum === 1) {
        const tabsOuterContainer = document.getElementsByClassName(
          'carousel-tabs-inner'
        )[0];
        const tabsOuterContainerRightPos = tabsOuterContainer.getBoundingClientRect()
          .right;
        const tabInnerContainerRightPos = tabsInnnerContainer.getBoundingClientRect()
          .right;
        const rightLength =
          tabInnerContainerRightPos - tabsOuterContainerRightPos;

        nextTabsPos = currentTabsPos - rightLength - 10;
      }

      tabsInnnerContainer.style.transform = `translateX(${nextTabsPos}px)`;

      this.setState({
        hiddenNextTabsNum: hiddenNextTabsNum - 1,
        hiddenPrevTabsNum: hiddenPrevTabsNum + 1,
        currentTabsPos: nextTabsPos,
      });
    }
  }

  showPrevTab() {
    const {
      activeIndex,
      hiddenNextTabsNum,
      hiddenPrevTabsNum,
      currentTabsPos,
      visibleTabsNum,
      itemsLength,
    } = this.state;

    // At showing hidden left tabs
    if (currentTabsPos < 0) {
      const tabsContainer = document.getElementsByClassName('carousel-tabs')[0];
      const tabItemWidth = document.getElementsByClassName('carousel-tab')[0]
        .clientWidth;
      const absCurrentTabsPos = Math.abs(currentTabsPos);
      let nextTabsPos;

      // At last tab
      if (activeIndex === itemsLength - 1) {
        nextTabsPos = -(tabItemWidth + 20);
      } else if (absCurrentTabsPos < tabItemWidth) {
        nextTabsPos = 0;
      } else {
        nextTabsPos = currentTabsPos + tabItemWidth;
      }

      tabsContainer.style.transform = `translateX(${nextTabsPos}px)`;

      this.setState({
        hiddenNextTabsNum: hiddenNextTabsNum + 1,
        hiddenPrevTabsNum: hiddenPrevTabsNum - 1,
        currentTabsPos: nextTabsPos,
      });

      if (activeIndex === 0) {
        this.setState({
          hiddenPrevTabsNum: 0,
          hiddenNextTabsNum: itemsLength - visibleTabsNum,
        });
      }
    }
  }

  showIndexTab(index) {
    const { items } = this.props;
    const {
      visibleTabsNum,
      hiddenPrevTabsNum,
      hiddenNextTabsNum,
      currentTabsPos,
    } = this.state;
    const tabsContainer = document.getElementsByClassName('carousel-tabs')[0];
    const tabItemWidth = document.getElementsByClassName('carousel-tab')[0]
      .clientWidth;
    let nextTabsPos;

    if (index > visibleTabsNum) {
      // If clicked indicator is larger than hidden next tabs
      nextTabsPos = -((index - visibleTabsNum) * tabItemWidth);

      this.setState({
        currentTabsPos: nextTabsPos,
        hiddenNextTabsNum: items.length - index,
        hiddenPrevTabsNum: index - visibleTabsNum,
      });
    } else if (index <= hiddenPrevTabsNum) {
      // If clicked indicator is less than equal hidden prev tabs
      const isPrevOne = index === hiddenPrevTabsNum;

      nextTabsPos =
        isPrevOne && hiddenPrevTabsNum > 0
          ? currentTabsPos + tabItemWidth
          : -(index * tabItemWidth);

      this.setState({
        currentTabsPos: nextTabsPos,
        hiddenNextTabsNum: isPrevOne
          ? hiddenNextTabsNum - 1
          : items.length - visibleTabsNum - index,
        hiddenPrevTabsNum: index,
      });
    }

    tabsContainer.style.transform = `translateX(${nextTabsPos}px)`;
  }

  goToIndex(index) {
    const { swiper } = this.state;

    if (swiper) {
      swiper.slideTo(index);
      this.setState({
        activeIndex: index,
      });
    }
  }

  render() {
    const { activeIndex, itemsLength, isTabActive } = this.state;
    const { items } = this.props;
    const isFirstSlide = activeIndex === 0;
    const isLastSlide = activeIndex === itemsLength - 1;
    const isMobile = window.innerWidth < 576;
    const params = {
      modules: [SwiperModule.Pagination, SwiperModule.Navigation],
      pagination: {
        el: '.swiper-pagination',
        type: 'bullets',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next.rex-icon.chevron-right',
        prevEl: '.swiper-button-prev.rex-icon.chevron-left',
      },
      slidesPerView: isMobile ? 1 : 3,
      spaceBetween: isMobile ? 0 : 4,
      shouldSwiperUpdate: true,
      on: {
        slideChange: this.onSlideChange,
      },
    };

    const renderItems = () =>
      items.map((item, index) => (
        <CarouselItem
          key={item.id}
          id={item.id}
          itemId={index}
          title={item.title}
          caption={item.caption}
          src={item.src}
        />
      ));

    const setSwiper = swiperInstance => {
      this.setState({
        swiper: swiperInstance,
      });
    };

    return (
      <div className="swiper-main-container">
        {isTabActive && (
          <CarouselTabs
            items={items}
            onClick={this.goToIndex}
            activeIndex={activeIndex}
            isFirstSlide={isFirstSlide}
            isLastSlide={isLastSlide}
          />
        )}
        <Swiper
          {...params}
          getSwiper={swiperInstance => setSwiper(swiperInstance)}
          modules={[SwiperModule.Navigation, SwiperModule.Pagination]}
        >
          {renderItems(items)}
        </Swiper>
      </div>
    );
  }
}

Carousel.defaultProps = {
  items: node,
  goToIndex: () => {},
  isTabActive: false,
};

Carousel.propTypes = {
  items: arrayOf(
    shape({
      id: number,
      src: string,
      title: string,
      caption: string,
      tabName: string,
    })
  ),
  goToIndex: func,
  isTabActive: bool,
};

export default Carousel;
