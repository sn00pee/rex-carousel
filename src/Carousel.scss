.rex-swiper {
  h1 {
    text-decoration: underline;
  }
}

html, body {
  position: relative;
  height: 100%;
}
body {
  font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
  font-size: 14px;
  color:#000;
  margin: 0;
  padding: 0;
}

.swiper-container {
  width: 100%;
  height: 100%;
}

.swiper-slide {
  min-height: 200px;
  text-align: center;
  font-size: 18px;
  background: #ccc;
  /* Center slide text vertically */
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-align-items: center;
  align-items: center;

  background-repeat: no-repeat;
  background-size: cover;
  width: 376px !important;
  height: 200px;
  border-radius: 4px;
  display: inline-block;
  margin-right: 4px;
  margin-left: 4px;
  box-shadow: 0 0 0 0 rgba(0, 0, 0, 0);
  transition: box-shadow .2s ease-in-out;
  cursor: pointer;

  @media (max-width: 576px) {
    width: 100vw !important;
    margin: 0 !important;
  }
}

.swiper-main-container {
    width: 1150px;

    @media (max-width: 576px) {
      width: 100vw !important;
    }
}

.swiper-button-next,
.swiper-container-rtl,
.swiper-button-prev {
  background-image: none !important;
  width: 30px !important;
  height: 30px !important;
  background-color: #fff !important;
  border-radius: 15px;
  display: inline-block;
  width: 20px;
  height: 20px;
  background: no-repeat 50% / 100% 100%;
  position: absolute !important;
  top: calc(50% - 6px) !important;
  transform: translateY(-50%);
  margin-top: 0 !important;

  @media (max-width: 576px) {
    display: none !important;
  }
}

.rex-icon::before {
  color: red;
  font-size: 2rem;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.swiper-button-prev::before {
  left: 48%;
}

.swiper-button-nect::before {
  left: 52%;
}

.swiper-pagination-bullet {
  background-color: #bf0000 !important;
  opacity: 1 !important;
  width: 8px !important;
  height: 8px !important;
  border-radius: 15px;
  box-sizing: border-box !important;
  background-clip: border-box !important;
  border-bottom: none !important;
  border-top: none !important;
  position: relative;
  box-sizing: content-box;
  flex: 0 1 auto;
  width: 30px;
  height: 3px;
  margin-right: 3px;
  margin-left: 3px;
  text-indent: -999px;
  cursor: pointer;
  background-color: #fff;
  background-clip: padding-box;
  border-top: 10px solid transparent;
  border-bottom: 10px solid transparent;
  opacity: .5;
  transition: opacity 0.6s ease;

  &::after {
    width: 6px;
    height: 6px;
    background-color: #fff;
    position: absolute;
    content: '';
    top: 1px;
    left: 1px;
    border-radius: 3px;
  }
}

.swiper-pagination-bullet-active {
  &::after {
    content: none;
  }
}

.carousel-item-description {
  position: absolute;
  left: 2.4rem;
  bottom: 1.6rem;
  color: #fff;
  text-align: left;
}

.carousel-item-title {
  font-size: 1.6rem;
  font-weight: bold;
}

.carousel-item-caption {
  font-size: 1.2rem; 
}

/* Tabs */
.carousel-control-icon {
  width: 30px !important;
  height: 30px !important;
  background-color: #fff !important;
  border-radius: 15px;
}

.carousel-control-anchor {
  opacity: 1 !important;
  width: 80px !important;

  &.hide {
    display: none;
  }
}

.carousel-control-next-icon::before {
  left: 52%;
}

.carousel-control-prev-icon::before {
  left: 48%;
}

.carousel-tab {
  list-style: none;
  padding-left: 28px;
  padding-right: 28px;
  color: #9c9c9c;
  cursor: pointer;
  position: relative;
  min-width: 110px;
  text-align: center;

  &.active {
    color: #bf0000;

    &::before {
      content: '';
      width: 100%;
      left: 50%;
      height: 1px;
      background-color: #bf0000;
      display: block;
      top: -10px;
      position: absolute;
      transform: translateX(-50%);
    }
  }
}


.carousel-tabs {
  padding-left: unset;
  margin-bottom: 0;
  display: flex;
  transition: transform .3s ease-in;
  padding-top: 10px;
  padding-bottom: 10px;

  &-container {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 20px;
    border-top: 1px solid #d1d1d1;
    position: relative;

    @media (max-width: 576px) {
      display: none;
    }
  }

  &-inner {
    overflow: hidden;
    margin-left: 30px;
    margin-right: 30px;
    width: 100%;
    position: relative;

    &::after {
      content: '';
      width: 60px;
      height: 24px;
      position: absolute;
      right: -30px;
      top: 10px;
      background: linear-gradient(90deg, rgba(255,255,255,1) 0%, rgba(255,255,255,0) 100%);
    }
  }
}

.carousel-tabs-controller {
  line-height: 1;
  cursor: pointer;
  position: absolute;
  z-index: 9;
  top: 10px;

  .rex-icon {
    width: 20px;
    height: 20px;
  }

  &.hide {
    display: none;
  }
}

.carousel-tabs-left-controller {
    left: 0;
}

.carousel-tabs-right-controller {
    right: 0;
}

.swiper-pagination {
  position: unset !important;
}
