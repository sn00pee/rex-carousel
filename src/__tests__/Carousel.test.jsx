import { shallow, mount } from 'enzyme';
import React from 'react';
import Carousel from '../Carousel';

const items = [
  {
    id: 0,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item A',
  },
  {
    id: 1,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item B',
  },
  {
    id: 2,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item C',
  },
  {
    id: 3,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item D',
  },
  {
    id: 4,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item E',
  },
  {
    id: 5,
    src: 'https://picsum.photos/376/200',
    title: 'Sagano Bamboo Forest',
    caption: 'Located in Kyoto, Japan',
    tabName: 'Item F',
  },
];

describe('Carousel component', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('should have same length of items', () => {
    const wrapper = mount(<Carousel items={items} />);

    expect(wrapper.find('.swiper-wrapper').children()).toHaveLength(
      items.length
    );
  });

  it('renders a carousel item', () => {
    const wrapper = shallow(<Carousel items={items} />);

    expect(wrapper.find('.swiper-slide')).toBeDefined();
    expect(wrapper.find('.carousel-item-description')).toBeDefined();
    expect(wrapper.find('.carousel-item-title')).toBeDefined();
    expect(wrapper.find('.carousel-item-caption')).toBeDefined();
  });

  it('returns to match snapshot', () => {
    const wrapper = shallow(<Carousel items={items} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("doesn't break without items", () => {
    const wrapper = shallow(<Carousel items={[]} />);

    expect(wrapper.find('.swiper-wrapper')).toHaveLength(0);
  });

  it('renders carousel item when item is added', () => {
    const wrapper = mount(<Carousel items={[]} />);

    wrapper.setProps({
      items: [
        {
          id: 0,
          src: 'https://picsum.photos/376/200',
          title: 'Sagano Bamboo Forest',
          caption: 'Located in Kyoto, Japan',
          tabName: 'Item A',
        },
      ],
    });

    expect(wrapper.find('.carousel-item-title').text()).toEqual(
      'Sagano Bamboo Forest'
    );
  });

  it('has tabs if isTabActive prop is passed', () => {
    const wrapper = shallow(<Carousel items={items} isTabActive />);

    expect(wrapper.find('.carousel-tabs-container')).toBeDefined();
  });

  it('does not have tabs if isTabActive props is not passed', () => {
    const wrapper = shallow(<Carousel items={items} />);

    expect(wrapper.find('.carousel-tabs-container')).not.toBe(undefined);
  });
});

describe('CarouselTabs component', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('renders a carouselTabs item', () => {
    const wrapper = shallow(<Carousel items={items} isTabActive />);

    expect(wrapper.find('.carousel-tabs-container')).toBeDefined();
    expect(wrapper.find('.carousel-tabs-prev')).toBeDefined();
    expect(wrapper.find('.carousel-tabs-next')).toBeDefined();
    expect(wrapper.find('.carousel-tabs-inner')).toBeDefined();
  });

  it('has length of items minus 2', () => {
    const wrapper = shallow(<Carousel items={items} isTabActive />);

    expect(
      wrapper
        .find('CarouselTabs')
        .render()
        .find('.carousel-tabs')
        .children()
    ).toHaveLength(items.length - 2);
  });

  it('should call the onClick function when tab is clicked', () => {
    const wrapper = shallow(<Carousel items={items} isTabActive />);
    const mockedHandleClickTab = jest.fn();

    wrapper.instance().goToIndex = mockedHandleClickTab;
    wrapper.instance().goToIndex();
    expect(mockedHandleClickTab).toHaveBeenCalledTimes(1);
  });
});

