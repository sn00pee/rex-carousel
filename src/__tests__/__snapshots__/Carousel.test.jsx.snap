// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component returns to match snapshot 1`] = `ShallowWrapper {}`;
